app/scripts/controllers/permissions/index.js

```diff
@@ -0,0 +1,418 @@
+      this.permissions, { origin }
+    ))
+    return asMiddleware(engine)
+  }
+
+  /**
+   * Returns the accounts that should be exposed for the given origin domain,
+   * if any. This method exists for when a trusted context needs to know
+   * which accounts are exposed to a given domain.
+   *
+   * @param {string} origin - The origin string.
+   */
+  getAccounts (origin) {
+    return new Promise((resolve, _) => {
+
+      const req = { method: 'eth_accounts' }
+      const res = {}
+      this.permissions.providerMiddlewareFunction(
+        { origin }, req, res, () => {}, _end
+      )
+
+      function _end () {
+        if (res.error || !Array.isArray(res.result)) {
+          resolve([])
+        } else {
+          resolve(res.result)
+        }
+      }
+    })
+  }
+
+  /**
+   * Submits a permissions request to rpc-cap. Internal use only.
+   *
+   * @param {string} origin - The origin string.
+   * @param {IRequestedPermissions} permissions - The requested permissions.
+   */
+  _requestPermissions (origin, permissions) {
+    return new Promise((resolve, reject) => {
+
+      const req = { method: 'wallet_requestPermissions', params: [permissions] }
+      const res = {}
+      this.permissions.providerMiddlewareFunction(
+        { origin }, req, res, () => {}, _end
+      )
+
+      function _end (err) {
+        if (err || res.error) {
+          reject(err || res.error)
+        } else {
+          resolve(res.result)
+        }
+      }
+    })
+  }
+
+  /**
+   * User approval callback. The request can fail if the request is invalid.
+   *
+   * @param {object} approved the approved request object
+   */
+  async approvePermissionsRequest (approved, accounts) {
+
+    const { id } = approved.metadata
+    const approval = this.pendingApprovals[id]
+
+    try {
+
+      // attempt to finalize the request and resolve it
+      await this.finalizePermissionsRequest(approved.permissions, accounts)
+      approval.resolve(approved.permissions)
+
+    } catch (err) {
+
+      // if finalization fails, reject the request
+      approval.reject(ethErrors.rpc.invalidRequest({
+        message: err.message, data: err,
+      }))
+    }
+
+    delete this.pendingApprovals[id]
+  }
+
+  /**
+   * User rejection callback.
+   *
+   * @param {string} id the id of the rejected request
+   */
+  async rejectPermissionsRequest (id) {
+    const approval = this.pendingApprovals[id]
+    approval.reject(ethErrors.provider.userRejectedRequest())
+    delete this.pendingApprovals[id]
+  }
+
+  /**
+   * Grants the given origin the eth_accounts permission for the given account(s).
+   * This method should ONLY be called as a result of direct user action in the UI,
+   * with the intention of supporting legacy dapps that don't support EIP 1102.
+   *
+   * @param {string} origin - The origin to expose the account(s) to.
+   * @param {Array<string>} accounts - The account(s) to expose.
+   */
+  async legacyExposeAccounts (origin, accounts) {
```

Comment:
This function appears to be unused

Reply:
Huh. I think we're missing a UI feature: the ability to manually connect to a site. That was possible with the old Connections setting page. That's required to support pre-EIP-1102 sites, as this comment describes.

Reply:
This will be used with the merging of https://github.com/MetaMask/metamask-extension/pull/7593